### PR TITLE
612: javax.crypto.AEADBadTagException crashes in Keystore.decryptBytes

### LINKED ISSUE #612: javax.crypto.AEADBadTagException crashes in Keystore.decryptBytes
Two crashes reported here which seem very related? Currently affect about ~dozen people.

See also #165 

```
java.lang.RuntimeException: 
  at android.app.ActivityThread.handleBindApplication (ActivityThread.java:5942)
  at android.app.ActivityThread.access$1100 (ActivityThread.java:202)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1658)
  at android.os.Handler.dispatchMessage (Handler.java:106)
  at android.os.Looper.loop (Looper.java:193)
  at android.app.ActivityThread.main (ActivityThread.java:6744)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:495)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:859)
Caused by: javax.crypto.AEADBadTagException: 
  at android.security.keystore.AndroidKeyStoreCipherSpiBase.engineDoFinal (AndroidKeyStoreCipherSpiBase.java:517)
  at javax.crypto.Cipher.doFinal (Cipher.java:2055)
  at mozilla.components.lib.dataprotect.Keystore.decryptBytes (Keystore.kt:244)
  at mozilla.lockbox.support.SecurePreferences.getString (SecurePreferences.kt:40)
  at mozilla.lockbox.store.AccountStore.getStoredAccountJSON (AccountStore.kt:76)
  at mozilla.lockbox.store.AccountStore.detectAccount (AccountStore.kt:120)
  at mozilla.lockbox.store.AccountStore.injectContext (AccountStore.kt:116)
  at mozilla.lockbox.LockboxApplication.injectContext (LockboxApplication.kt:108)
  at mozilla.lockbox.LockboxApplication.onCreate (LockboxApplication.kt:59)
  at android.app.Instrumentation.callApplicationOnCreate (Instrumentation.java:1155)
  at android.app.ActivityThread.handleBindApplication (ActivityThread.java:5937)
Caused by: android.security.KeyStoreException: 
  at android.security.KeyStore.getKeyStoreException (KeyStore.java:905)
  at android.security.keystore.KeyStoreCryptoOperationChunkedStreamer.doFinal (KeyStoreCryptoOperationChunkedStreamer.java:224)
  at android.security.keystore.AndroidKeyStoreAuthenticatedAESCipherSpi$BufferAllOutputUntilDoFinalStreamer.doFinal (AndroidKeyStoreAuthenticatedAESCipherSpi.java:373)
  at android.security.keystore.AndroidKeyStoreCipherSpiBase.engineDoFinal (AndroidKeyStoreCipherSpiBase.java:506)
```

```
java.lang.RuntimeException: 
  at android.app.ActivityThread.handleBindApplication (ActivityThread.java:5743)
  at de.robv.android.xposed.XposedBridge.invokeOriginalMethodNative (Native Method)
  at de.robv.android.xposed.XposedBridge.handleHookedMethod (XposedBridge.java:361)
  at android.app.ActivityThread.handleBindApplication (<Xposed>)
  at android.app.ActivityThread.-wrap1 (Unknown Source)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1656)
  at android.os.Handler.dispatchMessage (Handler.java:106)
  at android.os.Looper.loop (Looper.java:164)
  at android.app.ActivityThread.main (ActivityThread.java:6494)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:440)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:807)
  at de.robv.android.xposed.XposedBridge.main (XposedBridge.java:108)
Caused by: javax.crypto.AEADBadTagException: 
  at android.security.keystore.AndroidKeyStoreCipherSpiBase.engineDoFinal (AndroidKeyStoreCipherSpiBase.java:517)
  at javax.crypto.Cipher.doFinal (Cipher.java:1741)
  at mozilla.components.lib.dataprotect.Keystore.decryptBytes (Keystore.kt:244)
  at mozilla.lockbox.support.SecurePreferences.getString (SecurePreferences.kt:40)
  at mozilla.lockbox.store.AccountStore.getStoredAccountJSON (AccountStore.kt:76)
  at mozilla.lockbox.store.AccountStore.detectAccount (AccountStore.kt:120)
  at mozilla.lockbox.store.AccountStore.injectContext (AccountStore.kt:116)
  at mozilla.lockbox.LockboxApplication.injectContext (LockboxApplication.kt:108)
  at mozilla.lockbox.LockboxApplication.onCreate (LockboxApplication.kt:59)
  at android.app.Instrumentation.callApplicationOnCreate (Instrumentation.java:1119)
  at android.app.ActivityThread.handleBindApplication (ActivityThread.java:5740)
Caused by: android.security.KeyStoreException: 
  at android.security.KeyStore.getKeyStoreException (KeyStore.java:697)
  at android.security.keystore.KeyStoreCryptoOperationChunkedStreamer.doFinal (KeyStoreCryptoOperationChunkedStreamer.java:224)
  at android.security.keystore.AndroidKeyStoreAuthenticatedAESCipherSpi$BufferAllOutputUntilDoFinalStreamer.doFinal (AndroidKeyStoreAuthenticatedAESCipherSpi.java:373)
  at android.security.keystore.AndroidKeyStoreCipherSpiBase.engineDoFinal (AndroidKeyStoreCipherSpiBase.java:506)
```

### PR DESCRIPTION
**Pull request by eliserichards**
Fixes #612

## To Do

- [x] double check the original issue to confirm it is fully satisfied
- [ ] add testing notes and screenshots in PR description to help guide reviewers
- [ ] make sure CI builds are passing (e.g.: fix lint and other errors)

### CODE REVIEW

File: app/src/main/java/mozilla/lockbox/support/SecurePreferences.kt [hunk 1]
```diff
@@ -11,6 +11,7 @@ import android.content.SharedPreferences
 import android.preference.PreferenceManager
 import android.util.Base64
 import mozilla.components.lib.dataprotect.Keystore
+import mozilla.lockbox.LogProvider.Companion.log
```
Comment (jhugman) [inline app/src/main/java/mozilla/lockbox/support/SecurePreferences.kt]:
```kotlin
import mozilla.lockbox.log
```

File: app/src/main/java/mozilla/lockbox/support/SecurePreferences.kt [hunk 2]
```diff
@@ -39,7 +40,8 @@ open class SecurePreferences(
             try {
                 val plain = keystore.decryptBytes(encrypted)
                 String(plain, StandardCharsets.UTF_8)
-            } catch (error: IllegalArgumentException) {
+            } catch (error: Throwable) {
+                log.error("DecryptBytes exception: ", error)
```
Comment (jhugman) [inline app/src/main/java/mozilla/lockbox/support/SecurePreferences.kt]: There are quite a few exceptions possible from this method: would it make sense to be a bit more specific, using multiple catch blocks?

e.g. Some of these may be recoverable, some may be our fault, some may not.

File: app/src/main/java/mozilla/lockbox/store/AccountStore.kt [hunk 3]
```diff
@@ -38,9 +38,10 @@ import mozilla.lockbox.support.Constant
 import mozilla.lockbox.support.Optional
 import mozilla.lockbox.support.SecurePreferences
 import mozilla.lockbox.support.asOptional
+import java.lang.Exception
 import java.util.concurrent.TimeUnit
 import kotlin.coroutines.CoroutineContext
-import org.mozilla.fxaclient.internal.FxaException as FxaException
+import org.mozilla.fxaclient.internal.FxaException
```
Comment (jhugman) [inline app/src/main/java/mozilla/lockbox/store/AccountStore.kt]: Is this still needed now?

Comment (eliserichards) [reply app/src/main/java/mozilla/lockbox/store/AccountStore.kt]: Yep! it's still used, it's just not complaining about the alias anymore

File: app/src/main/java/mozilla/lockbox/support/SecurePreferences.kt [hunk 4]
```diff
@@ -40,6 +42,11 @@ open class SecurePreferences(
                 val plain = keystore.decryptBytes(encrypted)
                 String(plain, StandardCharsets.UTF_8)
             } catch (error: IllegalArgumentException) {
+                log.error("IllegalArgumentException exception: ", error)
+                null
+            } catch (error: AEADBadTagException) {
```
Comment (jhugman) [inline app/src/main/java/mozilla/lockbox/support/SecurePreferences.kt]: I think I'd've used a [GeneralSecurityException][1] here. 

[1]: https://docs.oracle.com/javase/8/docs/api/java/security/GeneralSecurityException.html

File: app/src/main/java/mozilla/lockbox/store/AccountStore.kt [hunk 5]
```diff
@@ -178,7 +179,7 @@ open class AccountStore(
             val config = Config.release(Constant.FxA.clientID, Constant.FxA.redirectUri)
             fxa = FirefoxAccount(config)
             generateLoginURL()
-        } catch (e: FxaException) {
+        } catch (e: Exception) {
```
Comment (jhugman) [inline app/src/main/java/mozilla/lockbox/store/AccountStore.kt]: See my previous comments about catching more specific errors.

File: app/src/main/java/mozilla/lockbox/store/AccountStore.kt [hunk 6]
```diff
@@ -123,7 +124,7 @@ open class AccountStore(
             } else {
                 try {
                     this.fxa = FirefoxAccount.fromJSONString(accountJSON)
-                } catch (e: FxaException) {
+                } catch (e: Exception) {
```
Comment (jhugman) [inline app/src/main/java/mozilla/lockbox/store/AccountStore.kt]: See my previous comments about catching more specific errors.

File: app/src/main/java/mozilla/lockbox/support/FxASyncDataStoreSupport.kt [hunk 7]
```diff
@@ -15,7 +15,7 @@ import mozilla.lockbox.store.ContextStore
 import java.security.SecureRandom
 
 /**
- * @param keyStrength The strength ofthe generated key in bits
+ * @param keyStrength The strength of the generated key in bits
```
Comment (linuxwolf) [inline app/src/main/java/mozilla/lockbox/support/FxASyncDataStoreSupport.kt]: thanks for fixing my typos 🙃

Comment (eliserichards) [reply app/src/main/java/mozilla/lockbox/support/FxASyncDataStoreSupport.kt]: No problem 😝

**Pull request merged**